### DISCUSSION TITLE
unable to uninstall xfrm policy after stopping strongswan

### DISCUSSION THREAD

**Discussion by SoldierProgrammer** [strongswan/strongswan#989] [Q&A]
Linux: CentOS 7
Strongswan: 5.5.1

Config:
conn xxx
    left=a.b.c.d
    right=aa.bb.cc.dd
    rightid=%any
    ike=aes128-sha1-modp3072!
    esp=aes128-sha1!
    ikelifetime=86400
    lifetime=3600
    keyexchange=ikev1
    type=tunnel
    authby=psk
    dpdaction=hold
    dpddelay=15
    dpdtimeout=90
    closeaction=none
    auto=ignore

conn xxx_1
    leftsubnet=10.60.3.0/24
    rightsubnet=192.168.1.0/24
    also=xxx
    auto=route
conn xxx_2
    leftsubnet=10.60.2.0/24
    rightsubnet=192.168.1.0/24
    also=xxx
    auto=route
conn xxx_3
    leftsubnet=10.60.1.0/24
    rightsubnet=192.168.1.0/24
    also=xxx
    auto=route

I executed the command "ipsec stop" after successful negotiation

Then I check the xfrm policy  by the command "ip xfrm policy":

src 192.168.1.0/24 dst 10.60.1.0/24 
	dir fwd priority 287712 ptype main 
	tmpl src aa.bb.cc.dd dst a.b.c.d
		proto esp reqid 3 mode tunnel
src 192.168.1.0/24 dst 10.60.1.0/24 
	dir in priority 287712 ptype main 
	tmpl src aa.bb.cc.dd dst a.b.c.d
		proto esp reqid 3 mode tunnel
src 10.60.1.0/24 dst 192.168.1.0/24 
	dir out priority 287712 ptype main 
	tmpl src a.b.c.d dst aa.bb.cc.dd
		proto esp reqid 3 mode tunnel

Why are there policy residues？？？

**Comment by tobiasbrunner** [comment]
[Increase the log level](https://docs.strongswan.org/strongswan-docs/5.9/config/logging.html) for _knl_ to 2 and read the log to see what's happening. You are using a very old version of strongSwan, so if it was a bug, it's most likely been fixed by now, so please try a more recent version. And it's even possible that a kernel bug could cause this.

**Reply by SoldierProgrammer** [reply]
No significant anomalies found. Log below:

2022-04-07 16:00:15.475 00[DMN] signal of type SIGINT received. Shutting down
2022-04-07 16:00:15.475 00[MGR] going to destroy IKE_SA manager and all managed IKE_SA's
2022-04-07 16:00:15.475 00[MGR] set driveout flags for all stored IKE_SA's
2022-04-07 16:00:15.475 00[MGR] wait for all threads to leave IKE_SA's
2022-04-07 16:00:15.475 00[MGR] delete all IKE_SA's
2022-04-07 16:00:15.475 00[MGR] destroy all entries
2022-04-07 16:00:15.475 00[KNL] deleting policy 10.60.1.0/24 === 192.168.0.0/16 out
2022-04-07 16:00:15.475 00[KNL] deleting policy 192.168.0.0/16 === 10.60.1.0/24 in
2022-04-07 16:00:15.475 00[KNL] deleting policy 192.168.0.0/16 === 10.60.1.0/24 fwd
2022-04-07 16:00:15.475 07[JOB] terminated worker thread 07
2022-04-07 16:00:15.475 01[JOB] terminated worker thread 01
2022-04-07 16:00:15.475 03[JOB] terminated worker thread 03
2022-04-07 16:00:15.475 05[JOB] terminated worker thread 05

Add some information below：
1、ipsec start，just after starting the strongswan:
src 192.168.0.0/16 dst 10.60.1.0/24
dir fwd priority 289760 ptype main
tmpl src 106.75.230.195 dst 117.50.179.245
proto esp reqid 1 mode tunnel
src 192.168.0.0/16 dst 10.60.1.0/24
dir in priority 289760 ptype main
tmpl src 106.75.230.195 dst 117.50.179.245
proto esp reqid 1 mode tunnel
src 10.60.1.0/24 dst 192.168.0.0/16
dir out priority 289760 ptype main
tmpl src 117.50.179.245 dst 106.75.230.195
proto esp reqid 1 mode tunnel
2、after successful negotiation, there are double as beblow:
src 192.168.0.0/16 dst 10.60.1.0/24
dir fwd priority 189760 ptype main
tmpl src 106.75.230.195 dst 117.50.179.245
proto esp reqid 1 mode tunnel
src 192.168.0.0/16 dst 10.60.1.0/24
dir in priority 189760 ptype main
tmpl src 106.75.230.195 dst 117.50.179.245
proto esp reqid 1 mode tunnel
src 10.60.1.0/24 dst 192.168.0.0/16
dir out priority 189760 ptype main
tmpl src 117.50.179.245 dst 106.75.230.195
proto esp reqid 1 mode tunnel
src 192.168.0.0/16 dst 10.60.1.0/24
dir fwd priority 289760 ptype main
tmpl src 106.75.230.195 dst 117.50.179.245
proto esp reqid 1 mode tunnel
src 192.168.0.0/16 dst 10.60.1.0/24
dir in priority 289760 ptype main
tmpl src 106.75.230.195 dst 117.50.179.245
proto esp reqid 1 mode tunnel
src 10.60.1.0/24 dst 192.168.0.0/16
dir out priority 289760 ptype main
tmpl src 117.50.179.245 dst 106.75.230.195
proto esp reqid 1 mode tunnel

It's not likely to be a kernel problem, and There are many lower strongswan versions in our network that are unlikely to be upgraded to the latest version immediately.

**Reply by tobiasbrunner** [reply]
> It's not likely to be a kernel problem

What makes you say that? Because from what I can see above, it definitely looks like a kernel bug (duplicate policies with different priorities) that was fixed with [this commit](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=4f47e8ab6ab796b5380f74866fa5287aca4dcc58).

**Reply by SoldierProgrammer** [reply]
Because I have tried strongswan 5.9.1 in the same kernel, there are no duplicate policies problem in the xfrm.
Is it possible that StrongSwan fixed the problem？？

**Reply by tobiasbrunner** [reply]
Don't really see how, because the Linux kernel doesn't accept duplicate policies (i.e. what you see above can't happen without that kernel bug - or a similar one). Are you sure you used 5.9.1 with XFRM (_kernel-netlink_ plugin only) and not with _kernel-libipsec_?

**Reply by SoldierProgrammer** [reply]
I'm sure the strongswan version is 5.9.1:

[root@192-168-1-179 strongswan]# ipsec version 
Linux strongSwan U5.9.1/K4.18.0-240.1.1.el8_3.x86_64
University of Applied Sciences Rapperswil, Switzerland
See 'strongswan --copyright' for copyright information.

But how to check XFRM (kernel-netlink plugin only) and not with kernel-libipsec?

from the strongswan log, loaded plugins in 5.9.1 are as below, only kernel-netlink can be found, no kernel-libipsec, is it OK?

Jan  4 15:15:40 00[LIB] loaded plugins: charon pkcs11 tpm aesni aes des rc2 sha2 sha1 md4 md5 mgf1 random nonce x509 revocation constraints acert pubkey pkcs1 pkcs7 pkcs8 pkcs12 pgp dnskey sshkey pem openssl gcrypt fips-prf gmp curve25519 chapoly xcbc cmac hmac ctr ccm gcm drbg newhope curl attr kernel-netlink resolve socket-default farp stroke vici updown eap-identity eap-sim eap-aka eap-aka-3gpp eap-aka-3gpp2 eap-md5 eap-gtc eap-mschapv2 eap-dynamic eap-radius eap-tls eap-ttls eap-peap xauth-generic xauth-eap xauth-pam xauth-noauth dhcp led duplicheck unity counters

Add some information for 5.5.1
loaded plugins: charon aes des rc2 sha2 sha1 md5 random nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs8 pkcs12 pgp dnskey sshkey pem fips-prf gmp xcbc cmac hmac attr kernel-netlink resolve socket-default stroke vici updown xauth-generic

There is duplicheck plugin in 5.9.1 but not in 5.5.1, What is this plug-in used for？？？

**Reply by tobiasbrunner** [reply]
> only kernel-netlink can be found, no kernel-libipsec, is it OK?

Yes, looks fine but the date is weird, unless you really got that from a three month old log file (or the date on your system is wrong). And did you do the same test? Configure trap policies, successfully establish the SA, terminate the daemon (and check the policies in the kernel after each of these steps)?

> There is duplicheck plugin in 5.9.1 but not in 5.5.1, What is this plug-in used for？？？

You generally don't want to have it enabled because it deletes both IKE_SAs if an active duplicate is found (see [here](https://docs.strongswan.org/strongswan-docs/5.9/plugins/duplicheck.html) for more details).

I can only repeat what I wrote above. The kernel generally doesn't accept duplicate policies with the same selectors. So when the trap policy that's installed with `auto=route` is updated when the CHILD_SA is established, no duplicate policies should get created. Instead the existing policies with the same selectors should get updated with new priorities/templates. That is, what you saw above can only happen if the kernel incorrectly doesn't consider the policies as equal (due to the different priority) and instead of updating the existing policies creates new ones.

**Reply by SoldierProgrammer** [reply]
Please ignore the date, the strongswan 5.9.1 is installed  and started several months ago. 
I have recorded the operation records and information output in the two files attached (two versions of strongswan)
[5.9.1.txt](https://github.com/strongswan/strongswan/files/8442866/5.9.1.txt)
[5.5.1.txt](https://github.com/strongswan/strongswan/files/8442867/5.5.1.txt)

**Reply by tobiasbrunner** [reply]
And you are claiming these two hosts (`hb13-unet-12-9` and `192-168-1-179`) run the same exact kernel (version/patch level)?

**Reply by SoldierProgrammer** [reply]
192-168-1-179  is a VM, the kernel is 4.18.    hb13-unet-12-9 is a physical machine, the kernal is 4.14

I found the duplicated policies  problem is fixed in kernel 5.7    
So I  thought there would be problems with both versions 4.18 and 4.14.

Now I think of a possibility that the virtual machine kernel may have been fixed, although the version is very low, I will check with the technician tomorrow. And check the the patch you mentioned.

Thank you very much.

**Reply by tobiasbrunner** [reply]
The problem is that distribution kernels (or stable kernels in general) are often heavily patched, so the version number does not tell you everything. The original commit that introduced the possibility of duplicate policies with different priorities ([7cb8a93968e3](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=7cb8a93968e3)) was already added with 3.9. But it wasn't an issue, in practice, due to the first check in `xfrm_policy_mark_match()`.

Until that check was removed with [ed17b8d377ea](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=ed17b8d377ea), which was first included in 5.7. Since this commit had a `Fixes:` line that referenced the patch from 3.9, it could have been backported to many stable and distribution kernels (for instance, it was backported to the [`linux-4.14.y`](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?id=1cd914b02b5ae999b04f44871f39dde4bffde96e) and [`linux-4.19.y`](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?id=3aa98483427942d4c65a1fc68b89c4a2763af479) branches, but not the `linux-4.18.y` branch, since 4.18 is not an official longterm kernel).

The [fix for it](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=4f47e8ab6ab796b5380f74866fa5287aca4dcc58) that was included in 5.8 did the same (besides ed17b8d377ea from 5.7 it also referenced an even older commit from 2.6.34, which might not have been 100% correct technically, it probably should have referenced 7cb8a93968e3 from 3.9 again). So it should have been backported to those older kernels too. It's included in [`linux-5.7.y`](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?id=7388c5d12dc5bd7efb2719878be4b0962b15282a), which introduced the problematic commit, and also in [`linux-5.4.y`](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?id=0307da6866606b8cfacda062c47cdf6588b6dfac), which was the latest longterm kernel at the time.

However, it's currently not contained in the `linux-4.14.y` or `linux-4.19.y` branches (and neither in `linux-4.18.y`, which is fine as it doesn't contain the problematic commit either).  Not sure if that was on purpose, or due to an oversight. The bottom line is that kernels based on those older stable trees (`linux-4.9.y` is also affected) have this issue with duplicate policies.

**Reply by SoldierProgrammer** [reply]
Thank you very much, you are so cool.
I get it now.

Another small question: 
Is it the normal process as below?

When StrongSwan starts, XFRM is told to create a policy
When SAs are established, XFRM is told to update the policy priority (without duplicate policies)
When SAs are deleted, XFRM is told to update the policy too
When StrongSwan stops, XFRM is told to delete the policy

**Reply by tobiasbrunner** [reply]
> When StrongSwan starts, XFRM is told to create a policy

Yes, but only if you configure a trap policy (`auto=route`). When matching traffic hits such a policy, the IKE daemon is instructed to initiate corresponding IKE and IPsec SAs.

> When SAs are established, XFRM is told to update the policy priority (without duplicate policies)

Yes, the priorities for "regular" policies are higher (lower numeric value) than for trap policies. In newer versions outbound policies also have the SPIs in their templates (there could be other changes too, e.g. the mode if transport mode was configured but the peer only supports tunnel mode).

> When SAs are deleted, XFRM is told to update the policy too

Yes, the original trap policy is restored.

> When StrongSwan stops, XFRM is told to delete the policy

Yes, except if you don't use trap policies, in which case they get removed already in the previous step.

**Reply by SoldierProgrammer** [reply]
You are so nice. Thanks again.
If I want to patch the 4.14 kernel, will this commit be enough?
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=4f47e8ab6ab796b5380f74866fa5287aca4dcc58

**Reply by tobiasbrunner** [reply]
> If I want to patch the 4.14 kernel, will this commit be enough?

Technically, yes. However, the patch does not apply cleanly (probably the reason why the fix was not automatically backported). I've prepared versions for [4.19](https://github.com/strongswan/strongswan/files/8470591/0001-xfrm-policy-match-with-both-mark-and-mask-on-user-in-4.19.y.txt), and [4.14/4.9](https://github.com/strongswan/strongswan/files/8470601/0001-xfrm-policy-match-with-both-mark-and-mask-on-user-in-4.9.y-4.14.y.txt) and submitted them to the stable kernel maintainers, so they will hopefully be included in future releases of these kernels.

**Reply by SoldierProgrammer** [reply]
Hi, tobiasbrunner
Now I use strongswan 5.9.1 and Centos Linux 4.14 kernel. There are some errors below:
2022-04-14 15:05:56.753 09[CFG] <vpntunnel|68> unable to install policy 192.168.100.0/24 === 10.9.0.0/16 in for reqid 2, the same policy for reqid 1 exists
2022-04-14 15:05:56.753 09[CFG] <vpntunnel|68> unable to install policy 192.168.100.0/24 === 10.9.0.0/16 fwd for reqid 2, the same policy for reqid 1 exists
2022-04-14 15:05:56.753 09[CFG] <vpntunnel|68> unable to install policy 10.9.0.0/16 === 192.168.100.0/24 out for reqid 2, the same policy for reqid 1 exists

Dose this is the xfrm duplicate policies problem??


The config is :
config setup
    uniqueids = yes

conn vpntunnel-test
    left=x.x.x.x
    right=y.y.y.y
    leftid="x.x.x.x"
    rightid="y.y.y.y"
    ike=aes192-sha1-modp1024!
    esp=aes192-sha1-modp1024!
    ikelifetime=86400
    lifetime=86400
    keyexchange=ikev2
    type=tunnel
    authby=psk
    dpdaction=hold
    dpddelay=15
    dpdtimeout=90
    closeaction=none
    auto=ignore

conn vpntunnel-test_1
    leftsubnet=10.9.0.0/16
    rightsubnet=100.104.0.0/16,192.168.100.0/24
    also=vpntunnel-test
    auto=route


xfrm policy is :
src 10.9.0.0/16 dst 192.168.100.0/24 
	dir out priority 379520 ptype main 
	tmpl src x.x.x.x dst y.y.y.y
		proto esp reqid 1 mode tunnel
src 192.168.100.0/24 dst 10.9.0.0/16 
	dir fwd priority 379520 ptype main 
	tmpl src y.y.y.y dst x.x.x.x
		proto esp reqid 1 mode tunnel
src 192.168.100.0/24 dst 10.9.0.0/16 
	dir in priority 379520 ptype main 
	tmpl src y.y.y.y dst x.x.x.x
		proto esp reqid 1 mode tunnel
src 10.9.0.0/16 dst 100.104.0.0/16 
	dir out priority 383616 ptype main 
	tmpl src x.x.x.x dst y.y.y.y
		proto esp reqid 1 mode tunnel
src 100.104.0.0/16 dst 10.9.0.0/16 
	dir fwd priority 383616 ptype main 
	tmpl src y.y.y.y dst x.x.x.x
		proto esp reqid 1 mode tunnel
src 100.104.0.0/16 dst 10.9.0.0/16 
	dir in priority 383616 ptype main 
	tmpl src y.y.y.y dst x.x.x.x
		proto esp reqid 1 mode tunnel

**Reply by tobiasbrunner** [reply]
No, that's something else. You configured a trap policy for two remote subnets. However, it seems that your peer only proposed a single subnet (you should see that in the log, maybe increase the log level for _cfg_ to 2). If it doesn't support multiple subnets per CHILD_SA, you have to configure separate configs for them.

**Reply by SoldierProgrammer** [reply]
Yes, you are right. If I configure two remote subnets into two separate conns, that's OK.

But there are still some questions:
1) The remote is Strongswan 5.5.1. It should support multiple subnets, but it only config one subnet above.
Dose it should work well? Maybe just establish one pair of child_sa

2) A pair of child sas can be negotiated successfully for traffic-triggered negotiation
The problem above is just triggered by command(ipsec up / swanctl -i)

3) I tried the strongswan 5.5.1 on both ends and it didn't have the above problems
Why??

**Reply by SoldierProgrammer** [reply]
2. A pair of child sas can be negotiated successfully for traffic-triggered negotiation
The problem above is just triggered by command(ipsec up / swanctl -i)
-------the flow is from this end to the remote is OK
            if the flow is from the remote to this end, still failed

**Reply by tobiasbrunner** [reply]
The problem is that reqids are allocated for the complete set of traffic selectors. If there is narrowing (e.g. to a single subnet) a new reqid gets allocated resulting in that error. If the SA is triggered by traffic (via acquire), the reqid from the trap policy is maintained. However, if you initiate it manually or the peer is the initiator that's not the case. In versions before 5.6.3 there was a reverse lookup to find a trap policy and get its reqid, that was removed because it was quite ugly and could have returned an incorrect value. To avoid this, configure the same subnets on both sides, or configure a static reqid in the config. The next release could also help as the error you got will be changed to a warning because there are situations where the reqid actually has to change.
